src/main/java/com/yelp/nrtsearch/server/luceneserver/NRTPrimaryNode.java

```diff
@@ -372,37 +389,39 @@ public void addReplica(int replicaID, ReplicationServerClient replicationServerC
       for (MergePreCopy preCopy : warmingSegments) {
         logger.debug("warming segment {}", preCopy.files.keySet());
         message("warming segment " + preCopy.files.keySet());
-        boolean found = false;
-        synchronized (preCopy.connections) {
-          for (ReplicationServerClient each : preCopy.connections) {
-            if (each.equals(replicationServerClient)) {
-              found = true;
-              break;
-            }
-          }
-        }
-
-        if (found) {
-          logMessage("this replica is already warming this segment; skipping");
+        if (preCopy.connections.contains(replicationServerClient)) {
+          logMessage(
+              String.format(
+                  "Replica %s:%d is already warming this segment",
+                  replicaDetails.replicationServerClient.getHost(),
+                  replicaDetails.replicationServerClient.getPort()));
           // It's possible (maybe) that the replica started up, then a merge kicked off, and it
           // warmed to this new replica, all before the
           // replica sent us this command:
           continue;
         }
 
-        // OK, this new replica is not already warming this segment, so attempt (could fail) to
-        // start warming now:
-        if (preCopy.tryAddConnection(replicationServerClient) == false) {
+        // Start copying the segments to the replica
+        FilesMetadata filesMetadata = RecvCopyStateHandler.writeFilesMetaData(preCopy.files);
+        Iterator<TransferStatus> transferStatusIterator =
+            replicationServerClient.copyFiles(indexName, primaryGen, filesMetadata);
+        logMessage(
+            String.format(
+                "Start precopying merged segments for new replica %s:%d",
+                replicaDetails.replicationServerClient.getHost(),
+                replicaDetails.replicationServerClient.getPort()));
+
+        // If the preCopy is still in progress add this transfer to it as well
+        if (!preCopy.tryAddConnection(replicationServerClient, transferStatusIterator)) {
```

Comment:
A small issue with this is that if the preCopy is already finished, the iterator never gets consumed. I think it might be better to do something like
```
synchronized(preCopy) {
  if (!preCopy.finished()) {
    // start copyFiles request
    // add request iterator to preCopy
  }
}
```